#ifndef MANAGERRESERVEWIDGET_H
#define MANAGERRESERVEWIDGET_H

#include <QtGui>
#include <QtSql>
#include "whitetable.h"
#include "greytable.h"
#include "constants.h"
#include "helpers.h"

namespace Ui {
class ManagerReserveWidget;
}

class ManagerReserveWidget : public QWidget
{
    Q_OBJECT
    
public:
    explicit ManagerReserveWidget(ReserveStruct rstruct,
                                  QWidget *parent = 0);
    ~ManagerReserveWidget();
    void set_grey_table(GreyTable *g);
    void set_reserve_header();
    int resize_section(int, int);

private:
    Ui::ManagerReserveWidget *ui;
    GreyTable *grey_table;
    QSqlQueryModel *source;
    QStandardItemModel *list_model;
    QItemSelectionModel *selection_model;
    QHeaderView *table_header;
    QSortFilterProxyModel *model;
    QMap<QString, QString> columns_names;
    void set_layout();
    void set_columns_names();
    void set_headers();
    void connects();

    int max_docs;
    QSortFilterProxyModel *oldmodel;
    void delpos(int);

    void fill_list();

signals:
    void need_refresh();
    void switch_reserves();

private slots:
    void change_doc(QModelIndex cur, QModelIndex prev = QModelIndex());
    void add_position();
    void subtr_quant();
    void delete_position();
    void view_doc();
    void restore_doc();
    void clear_doc();
    void move_doc();
    void refresh_table();
    double count_sum();
};


//////////////////////////////////////////////////////////////////////////////////////////////////////
/////////////////////////////////////////// –ö–õ–ê–°–° ///////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////
///////////////////////////////////// –î–û–ë–ê–í–ò–¢–¨ –ü–û–ó–ò–¶–ò–Æ ////////////////////////////////////////////
//////////////////////////////////////////////////////////////////////////////////////////////////

class ReserveAddPosition : public QDialog{
    Q_OBJECT

public:
    explicit ReserveAddPosition(int id, int num, int quantity, double retail_price, double whole_price, int whole_begin, QWidget *parent = 0);

private:
    // –∏–Ω—Ç–µ—Ä—Ñ–µ–π—Å
    QPushButton *accept_button;
    QPushButton *reject_button;
    QLineEdit *quantity_spin;
    QDoubleSpinBox *price;
    QLineEdit *note;
    //QLineEdit *base_price;
    QLineEdit *quantity_total;

    // —ç–ª–µ–º–µ–Ω—Ç—ã –∫–æ–Ω—Å—Ç—Ä—É–∫—Ç–æ—Ä–∞
    int grey_id, doc_num;
    int base_quantity;
    double base_retail_price, base_whole_price;
    int base_whole_begin;

    bool price_changed;
    bool editing;

    void set_layout();
    void set_base_price_text(double);


private slots:
    void price_entered();
    void quantity_changed(int);
    void accept();

};


//////////////////////////////////////////////////////////////////////////////////////////////////////
/////////////////////////////////////////// –ö–õ–ê–°–° ///////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////
//////////////////////////////////// –í–´–ß–ï–°–¢–¨ –ö–û–õ–ò–ß–ï–°–¢–í–û ///////////////////////////////////////////
//////////////////////////////////////////////////////////////////////////////////////////////////

class ReserveSubtract : public QDialog{
    Q_OBJECT

public:
    explicit ReserveSubtract(int id, int quantity, QWidget *parent = 0);

private:
    QSpinBox *subtr_spin;
    QLineEdit *quant_line;
    QPushButton *accept_button;
    QPushButton *reject_button;

    int c_id, c_num, c_quantity;

private slots:
    void change_quant(int);
    void accept();

};


//////////////////////////////////////////////////////////////////////////////////////////////////////
/////////////////////////////////////////// –ö–õ–ê–°–° ///////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////
//////////////////////////////////// –ü–ï– –ï–ú–ï–°–¢–ò–¢–¨ –ü–û–ó–ò–¶–ò–ò //////////////////////////////////////////
//////////////////////////////////////////////////////////////////////////////////////////////////

class ReserveMovePositions : public QDialog{
    Q_OBJECT

public:
    explicit ReserveMovePositions(QTableView *table, QModelIndexList indexes, QWidget *parent = 0);

private:
    // –≠–ª–µ–º–µ–Ω—Ç—ã –∏–Ω—Ç–µ—Ä—Ñ–µ–π—Å–∞:
    // - —Ç—Ä–∏ —Ä–∞–¥–∏–æ–∫–Ω–æ–ø–∫–∏ ("–∑–∞—è–≤–∫–∏", "–¥–æ–∫—É–º–µ–Ω—Ç—ã", "–∫–æ–º–º. –ø—Ä–µ–¥–ª."
    // - –∫–Ω–æ–ø–∫–∏ –û–ö –∏ –û—Ç–º–µ–Ω–∞
    // - QListView –¥–ª—è –∑–∞—è–≤–æ–∫, –¥–≤–∞ QTableView –¥–ª—è –¥–æ–∫—É–º–µ–Ω—Ç–æ–≤ –∏ –ø—Ä–µ–¥–ª–æ–∂–µ–Ω–∏–π
    // - QStackedLayout –¥–ª—è –∏—Ö –ø–µ—Ä–µ–∫–ª—é—á–µ–Ω–∏—è
    // - –¥–≤–∞ –≤–∏–¥–∂–µ—Ç–∞ - –∫–æ–Ω—Ç–µ–π–Ω–µ—Ä–∞ –¥–ª—è –¥–æ–∫—É–º–µ—Ç–æ–≤+—Ñ–∏–ª—å—Ç—Ä—ã –∏ –ø—Ä–µ–¥–ª–æ–∂–µ–Ω–∏–π+—Ñ–∏–ª—å—Ç—Ä—ã
    // - —Ñ–∏–ª—å—Ç—Ä—ã –¥–ª—è –¥–æ–∫—É–º–µ–Ω—Ç–æ–≤ –∏ –∫–æ–º–º. –ø—Ä–µ–¥–ª–æ–∂–µ–Ω–∏–π
    QRadioButton *reserves_button, *docs_button, *offers_button;
    QPushButton *ok_button, *cancel_button;
    QStackedLayout *layout;
    QWidget *docs_widget, *offers_widget;
    QListView *reserves_list;
    QTableView *docs_list, *offers_list;
    // –≤–∏–¥–∂–µ—Ç—ã —Ñ–∏–ª—å—Ç—Ä–æ–≤ –¥–æ–±–∞–≤–∏–º –ø–æ–∑–¥–Ω–µ–µ

    int max_docs;

    inline void create_widgets();
    inline void set_layout();
    inline void connects();
    inline void form_reserves_list();
    inline void form_docs_list();
    inline void form_offers_list();

private slots:
    void switch_widget();
};


#endif // MANAGERRESERVEWIDGET_H
